### ISSUE TITLE
Argument typehinting not enforced

### ISSUE THREAD

**Issue by jbenoit2011** [Microsoft/TypeScript#17593]
**TypeScript Version:**  2.4.2

**Code**

```ts
class A {}

class B {}

class Factory {
    create(foo: A) {
        console.log('foo is an instance of ' + foo.constructor);
    }
}

var f: Factory = new Factory();
f.create(new B());
```

**Expected behavior:**

Compilation and run failure since `create` method requires a `foo` parameter of type `A`.

**Actual behavior:**

Compilation and run with node succeed.

**Comment by RyanCavanaugh** [comment]
https://github.com/Microsoft/TypeScript/wiki/FAQ

**Comment by jbenoit2011** [comment]
Ok for the runtime type checking, I was not aware of that. I think it should be stated as well in the documentation, since IMHO more people read it than the FAQ.

But what about the compile-time type check which do not detect I pass an incorrect type?
With some extra experimentations, I noticed the compile time check successfully detect the error if class `A` is not empty.

**Comment by RyanCavanaugh** [comment]
https://github.com/Microsoft/TypeScript/wiki/FAQ#what-is-structural-typing
https://github.com/Microsoft/TypeScript/wiki/FAQ#why-do-these-empty-classes-behave-strangely
https://github.com/Microsoft/TypeScript/wiki/FAQ#when-and-why-are-classes-nominal
